### PR TITLE
Remote Control Info and Preview on Watch

### PR DESCRIPTION
**Pull request by ErAzOr2k**
Display data and preview from the remote control streamer instead of from the device connected to the watch. You have to enable this in the watch settings.

### CODE REVIEW

File: Common/Various/CommonUtils.swift [hunk 1]
```diff
@@ -521,6 +521,22 @@ func formatAudioLevelChannels(channels: Int) -> String {
...
+    } else if string.hasPrefix("Unknown,") {
+        return Float.infinity
+    } else if let range = string.range(of: " dB,") {
+        let levelString = String(string[string.startIndex ..< range.lowerBound])
+        return Float(levelString) ?? Float.nan
+    } else {
+        return Float.nan
+    }
+}
+
```
Comment (eerimoq) [inline Common/Various/CommonUtils.swift:539]: We should add the numeric value to the remote control protocol to simplify this and make it work with bars.

File: Moblin Watch/Various/Model.swift [hunk 2]
```diff
@@ -246,6 +246,11 @@ class Model: NSObject, ObservableObject {
         if self.settings.show == nil {
             self.settings.show = .init()
         }
+        if self.settings.remoteControl == true {
+            previewTimeout = Duration.seconds(5)
+        } else {
+            previewTimeout = Duration.seconds(10)
+        }
```
Comment (eerimoq) [inline Moblin Watch/Various/Model.swift:253]: why is this needed?

File: Moblin/Various/Model.swift [hunk 3]
```diff
@@ -4966,13 +4980,50 @@ extension Model {
...
+            }
+            if let bitrateMessage = remoteControlTopRight?.bitrate?.message {
+                sendSpeedAndTotalToWatch(speedAndTotal: bitrateMessage)
+            }
+            sendThermalStateToWatch(thermalState: thermalState)
+            sendAudioLevelToWatch(audioLevel: audioLevel)
+            sendIsLiveToWatch(isLive: remoteControlTopRight?.bitrate?.message != nil)
+            sendIsRecordingToWatch(isRecording: remoteControlTopRight?.recording?.message != nil)
+            sendIsMutedToWatch(isMuteOn: audioLevel.isNaN)
+        }
```
Comment (eerimoq) [inline Moblin/Various/Model.swift:5007]: Should this be sent when the remote control assistant receives updates, not send the previous update?

File: Moblin/Various/Model.swift [hunk 4]
```diff
@@ -5076,32 +5130,32 @@ extension Model {
         )
     }
 
-    private func sendSpeedAndTotalToWatch() {
+    private func sendSpeedAndTotalToWatch(speedAndTotal: Any) {
```
Comment (eerimoq) [inline Moblin/Various/Model.swift:5133]: Use `Any` as little as possible for lower risk of mistakes.

File: Moblin/Various/Model.swift [hunk 5]
```diff
@@ -5274,11 +5330,16 @@ extension Model: WCSessionDelegate {
...
+        if database.watch!.remoteControl! {
+            remoteControlAssistantSetStream(on: value)
+            sendIsLiveToWatch(isLive: value)
+        } else {
+            DispatchQueue.main.async {
+                if value {
+                    self.startStream()
+                } else {
+                    self.stopStream()
+                }
```
Comment (eerimoq) [inline Moblin/Various/Model.swift:5342]: The whole if-else should be in main async block, right?

File: Moblin/Various/Model.swift [hunk 6]
```diff
@@ -5287,11 +5348,16 @@ extension Model: WCSessionDelegate {
...
+        if database.watch!.remoteControl! {
+            remoteControlAssistantSetRecord(on: value)
+            sendIsRecordingToWatch(isRecording: value)
+        } else {
+            DispatchQueue.main.async {
+                if value {
+                    self.startRecording()
+                } else {
+                    self.stopRecording()
+                }
```
Comment (eerimoq) [inline Moblin/Various/Model.swift:5360]: Same comment as above.

File: Moblin/Various/Model.swift [hunk 7]
```diff
@@ -5300,8 +5366,13 @@ extension Model: WCSessionDelegate {
...
         }
-        DispatchQueue.main.async {
-            self.setIsMuted(value: value)
+        if database.watch!.remoteControl! {
+            remoteControlAssistantSetMute(on: value)
+            sendIsMutedToWatch(isMuteOn: value)
+        } else {
+            DispatchQueue.main.async {
+                self.setIsMuted(value: value)
+            }
```
Comment (eerimoq) [inline Moblin/Various/Model.swift:5375]: Same comment as above.

File: Moblin/View/Settings/Watch/WatchSettingsView.swift [hunk 8]
```diff
@@ -15,6 +15,19 @@ struct WatchSettingsView: View {
...
+                    model.database.watch!.remoteControl!
+                }, set: { value in
+                    model.database.watch!.remoteControl = value
+                    model.store()
+                }))
+            } footer: {
+                Text("""
+                Display data and preview from the remote control streamer \
+                instead of from the device connected to the watch.
+                """)
```
Comment (eerimoq) [inline Moblin/View/Settings/Watch/WatchSettingsView.swift:29]: I think this should say something like "Use the watch as remote control assistant instead of ...", or at least mention the assistant, as it is the assistant functionality we expose in the watch.

File: Common/Various/CommonUtils.swift (hunk 1 above)
Comment (ErAzOr2k) [reply Common/Various/CommonUtils.swift]: Still in progress

File: Moblin Watch/Various/Model.swift (hunk 2 above)
Comment (ErAzOr2k) [reply Moblin Watch/Various/Model.swift]: The interval for the clock is 5 seconds. For the remote assistant, it's 10 seconds. I have now also changed the interval for the remote assistant to 5 seconds. I hope that's okay.

File: Moblin/Various/Model.swift (hunk 3 above)
Comment (ErAzOr2k) [reply Moblin/Various/Model.swift:5009]: I don't quite understand. The listed conditions are necessary to set the toggles.

File: Moblin/Various/Model.swift (hunk 4 above)
Comment (ErAzOr2k) [reply Moblin/Various/Model.swift]: done

File: Moblin/Various/Model.swift (hunk 5 above)
Comment (ErAzOr2k) [reply Moblin/Various/Model.swift:5344]: done

File: Moblin/Various/Model.swift (hunk 6 above)
Comment (ErAzOr2k) [reply Moblin/Various/Model.swift:5362]: done

File: Moblin/Various/Model.swift (hunk 7 above)
Comment (ErAzOr2k) [reply Moblin/Various/Model.swift:5377]: done

File: Moblin/View/Settings/Watch/WatchSettingsView.swift (hunk 8 above)
Comment (ErAzOr2k) [reply Moblin/View/Settings/Watch/WatchSettingsView.swift:22]: I have written a new text

File: Common/Various/CommonUtils.swift (hunk 1 above)
Comment (ErAzOr2k) [reply Common/Various/CommonUtils.swift]: Done. Although I'm not entirely satisfied because the alignment doesn't quite fit (compared to all the other elements).

What do you think? should we recreate a new audiolevenview for remotecontrol?

Comment (eerimoq) [reply Common/Various/CommonUtils.swift]: We want the UI to be pretty. So I guess yes =)

Comment (eerimoq) [reply Common/Various/CommonUtils.swift]: Or possibly send both what we have today and the numeric level.

Comment (ErAzOr2k) [reply Common/Various/CommonUtils.swift]: ok much cleaner now :)

File: Moblin/Various/Model.swift (hunk 3 above)
Comment (eerimoq) [reply Moblin/Various/Model.swift:5009]: `updateRemoteControlAssistantStatus()` requests an update from the streamer. It's the completion blocks for `getStatus()` and `getSettings()` that receives the responses, and it is in those completion blocks we should update the watch.

File: Moblin/Various/Model.swift [hunk 9]
```diff
@@ -4955,7 +4970,7 @@ extension Model {
     }
 
     func updateRemoteControlAssistantStatus() {
-        guard showingRemoteControl, remoteControlAssistant?.isConnected() == true else {
```
Comment (eerimoq) [inline Moblin/Various/Model.swift:4958]: I would prefer the remote control assistant to only fetch information from the remote control streamer when the assistant is actually in use, to minimize the network traffic from the streamer.

Comment (ErAzOr2k) [reply Moblin/Various/Model.swift]: ok reverted it

File: Moblin/Various/Model.swift (hunk 3 above)
Comment (ErAzOr2k) [reply Moblin/Various/Model.swift]: hope it's okay now :)

File: Moblin/Various/Model.swift [hunk 10]
```diff
@@ -5288,10 +5351,15 @@ extension Model: WCSessionDelegate {
             return
         }
         DispatchQueue.main.async {
-            if value {
-                self.startRecording()
+            if self.isRemoteControlAssistantConnected() {
+                self.remoteControlAssistantSetRecord(on: value)
+                self.sendIsRecordingToWatch(isRecording: value)
```
Comment (eerimoq) [inline Moblin/Various/Model.swift:5356]: I can't remember the details, but if possible, we should not update the clock record state here, but instead when the streamer phone has responded and says it is recording.

File: Moblin/Various/Model.swift [hunk 11]
```diff
@@ -4962,6 +4977,32 @@ extension Model {
             self.remoteControlGeneral = general
             self.remoteControlTopLeft = topLeft
             self.remoteControlTopRight = topRight
+            if self.isRemoteControlAssistantConnected() {
```
Comment (eerimoq) [inline Moblin/Various/Model.swift:4980]: This check can be removed as the same check is done in the beginning of the function.

File: Moblin/Various/Model.swift (hunk 10 above)
Comment (ErAzOr2k) [reply Moblin/Various/Model.swift:5354]: betrifft das auch handleSetIsMuted und handleSetIsLive?

File: Moblin/Various/Model.swift (hunk 11 above)
Comment (ErAzOr2k) [reply Moblin/Various/Model.swift]: done

File: Moblin/Various/Model.swift (hunk 10 above)
Comment (eerimoq) [reply Moblin/Various/Model.swift:5354]: I think so, yes. But as I said, I can't remember is it's easy to do the the "correct" way. So don't put too much time into it, but give it a try.

Comment (ErAzOr2k) [reply Moblin/Various/Model.swift]: complete. I also fixed a serious bug with audio level

File: Moblin/Various/Model.swift [hunk 12]
```diff
@@ -4678,9 +4694,8 @@ extension Model: RemoteControlStreamerDelegate {
             }
             var topRight = RemoteControlStatusTopRight()
             if self.isShowingStatusAudioLevel() {
-                let level = formatAudioLevel(level: self.audioLevel) +
-                    formatAudioLevelChannels(channels: self.numberOfAudioChannels)
-                topRight.audioLevel = RemoteControlStatusItem(message: level)
+                topRight.audioLevel = self.audioLevel.isNaN ? nil : self.audioLevel
```
Comment (eerimoq) [inline Moblin/Various/Model.swift:4697]: What is the reason for assigning `self.audioLevel` conditionally?

File: Moblin/Various/Model.swift [hunk 13]
```diff
@@ -4955,13 +4970,37 @@ extension Model {
...
+            }
+            self.sendThermalStateToWatch(thermalState: thermalState)
+            if let recordingMessage = topRight.recording?.message {
+                self.sendRecordingLengthToWatch(recordingLength: recordingMessage)
+            }
+            self.sendIsRecordingToWatch(isRecording: topRight.recording?.message != nil)
+            if let bitrateMessage = topRight.bitrate?.message {
+                self.sendSpeedAndTotalToWatch(speedAndTotal: bitrateMessage)
+            }
+            self.sendIsLiveToWatch(isLive: topRight.bitrate?.message != nil)
```
Comment (eerimoq) [inline Moblin/Various/Model.swift:4999]: Why it `bitrate` used to check if live? Should we add the live state to the remote control protocol?

File: Moblin/Various/Model.swift [hunk 14]
```diff
@@ -4955,13 +4970,37 @@ extension Model {
...
+            if let audioLevel = topRight.audioLevel {
+                self.sendAudioLevelToWatch(audioLevel: audioLevel)
+            }
+            self.sendIsMutedToWatch(isMuteOn: topRight.audioLevel == nil)
```
Comment (eerimoq) [inline Moblin/Various/Model.swift:5003]: Isn't the audio level NaN when muted?

File: Moblin/Various/Model.swift (hunk 12 above)
Comment (ErAzOr2k) [reply Moblin/Various/Model.swift:4697]: The reason is, when audio is muted (audiolevel = .nan) updateRemoteControlAssistantStatus won't get called. The reason seems to be that nan is noch decodeable/encodeable. took me a while to figure it out.

File: Moblin/Various/Model.swift (hunk 14 above)
Comment (eerimoq) [reply Moblin/Various/Model.swift:5003]: Might even be possible yo remote isMuted fro mthe watch protocol now that we have audioLevel. It replaces it I think.

Comment (ErAzOr2k) [reply Moblin/Various/Model.swift:5003]: doesn't work with .nan

File: Moblin/Various/Model.swift (hunk 12 above)
Comment (eerimoq) [reply Moblin/Various/Model.swift:4697]: What about using an enum to represent the audio level? I have not tested it, but should work. And makes it more clear than using `nil`. Does infinity (as in unknown) work when encoding a float using JSON? Probably not?
``` swift
enum AudioLevel: Codable {
    case muted
    case unknown
    case value(Float)
}
```

Comment (ErAzOr2k) [reply Moblin/Various/Model.swift]: it works :)

File: Moblin/Various/Model.swift (hunk 13 above)
Comment (ErAzOr2k) [reply Moblin/Various/Model.swift]: done
